### PR TITLE
Feat pipeline version configration nf

### PR DESCRIPTION
**Pull request by nazarfil**
A short, meaningful PR description. Explain the goal of the PR and the ideas behind it.

## Changes

Adds config field to pipeline [version](https://bluesquare.atlassian.net/browse/HEXA-938)
Allows pipeline to run in scheduled mode if parameter is in config

### CODE REVIEW

File: hexa/pipelines/models.py [hunk 1]
```diff
@@ -262,6 +267,7 @@ def upload_new_version(
             description=description,
             external_link=external_link,
             zipfile=zipfile,
+            config=config if config else self.config,
```
Comment (qgerome) [inline hexa/pipelines/models.py:270]: `Pipeline.config` is only used for the former system of pipelines using Airflow. For our implementation we can ignore that.

File: hexa/pipelines/schema/mutations.py [hunk 2]
```diff
@@ -185,7 +185,7 @@ def resolve_run_pipeline(_, info, **kwargs):
             user=request.user,
             pipeline_version=version,
             trigger_mode=PipelineRunTrigger.MANUAL,
-            config=input.get("config", {}),
+            config=input.get("config", version.config),
```
Comment (qgerome) [inline hexa/pipelines/schema/mutations.py:188]: No need to use `version.config` as the default for the config. This logic has to be in the `run` method of the `PipelineVersion` in itself.

**Review by qgerome — CHANGES_REQUESTED** [review]
Already a great start. There is a behaviour missing in this PR though: When a user launches a pipeline run, we have to take the config provided via the graphql and merge it with the configuration on the pipeline version before running it. 

Also, I don't think the scheduler has been modifier to take into consideration that it can takes the config on the `PipelineVersion` to run the pipeline.

File: hexa/pipelines/schema/mutations.py (hunk 2 above)
Comment (nazarfil) [reply hexa/pipelines/schema/mutations.py]: Ok, changed it to merge both

File: hexa/pipelines/models.py (hunk 1 above)
Comment (qgerome) [reply hexa/pipelines/models.py:270]: Are you sure ? It's in the `upload_new_version` method that is on `Pipeline`

File: hexa/pipelines/models.py [hunk 3]
```diff
@@ -323,6 +328,15 @@ def get_token(self):
             }
         )
 
+    def merge_pipeline_config(
+        self,
+        config: typing.Mapping[typing.Dict, typing.Any],
+        pipeline_version_config: typing.Mapping[typing.Dict, typing.Any],
+    ):
+        provided_config = config if config else self.config
+        merged_config = {**pipeline_version_config, **provided_config}
```
Comment (qgerome) [inline hexa/pipelines/models.py:337]: I'm wondering if we should only takes the keys where their value is != than None?

File: hexa/pipelines/tests/test_schema/test_pipelines.py [hunk 4]
```diff
@@ -1471,6 +1471,55 @@ def test_pipelines_permissions_schedule_required_param_with_no_default_value(sel
             r["data"]["pipelineByCode"],
         )
 
+    def test_pipelines_permissions_schedule_required_param_with_no_default_value_with_version_config(
```
Comment (qgerome) [inline hexa/pipelines/tests/test_schema/test_pipelines.py:1474]: Could you write more tests to check how the config is merged when you run a pipeline with a version's config?

**Review by qgerome — CHANGES_REQUESTED** [review]

**Review by qgerome — CHANGES_REQUESTED** [review]
You have to check the pipelines_scheduler.py file. This PR will have an effect on the `Pipeline.is_schedulable` (since you can change a version's config to remove values that may be required by a pipeline to run using a schedule)

As of now, the scheduler takes all the pipelines that have a schedule set. It should take into consideration only the ones that can be scheduled

**Comment by nazarfil** [comment]
> You have to check the pipelines_scheduler.py file. This PR will have an effect on the `Pipeline.is_schedulable` (since you can change a version's config to remove values that may be required by a pipeline to run using a schedule)

Indeed, valid point. For the case where the pipeline was uploaded as schedulable, and if we remove a required parameter that was set previously in config - pipeline won't be schedulable anymore. I will add a check to forbid users uploading version config with required parameters that don't have default values and are removed.

**Comment by qgerome** [comment]
> > You have to check the pipelines_scheduler.py file. This PR will have an effect on the `Pipeline.is_schedulable` (since you can change a version's config to remove values that may be required by a pipeline to run using a schedule)
> 
> Indeed, valid point. For the case where the pipeline was uploaded as schedulable, and if we remove a required parameter that was set previously in config - pipeline won't be schedulable anymore. I will add a check to forbid users uploading version config with required parameters that don't have default values and are removed.

I'm not sure to understand what you are going to do, are you proposing to prevent users to remove the values of the version's config when the pipeline is scheduled and these parameters are marked as required in the configuration ?

**Comment by nazarfil** [comment]
> I'm not sure to understand what you are going to do, are you proposing to prevent users to remove the values of the version's config when the pipeline is scheduled and these parameters are marked as required in the configuration

Yes, preventing submission if a pipelineVersion that has config with a required parameter removed, when there is no default.

File: hexa/pipelines/models.py [hunk 5]
```diff
@@ -131,16 +133,31 @@ def update_if_has_perm(self, principal: User, **kwargs):
         if not principal.has_perm("pipelines.update_pipeline_version", self):
             raise PermissionDenied
 
-        for key in ["name", "description", "external_link"]:
+        for key in ["name", "description", "external_link", "config"]:
             if key in kwargs and kwargs[key] is not None:
+                if key == "config" and self.is_schedulable:
```
Comment (qgerome) [inline hexa/pipelines/models.py:138]: You can move that piece of code outside the for loop and treat the config directly.

Comment (qgerome) [reply hexa/pipelines/models.py:138]: `is_schedulable` does not mean that the pipeline has a `schedule` ⚠️ It's better to check `pipeline.schedule`.

**Review by qgerome — CHANGES_REQUESTED** [review]

**Pull request merged**